StRoot/StETofCalibMaker/StETofCalibMaker.cxx

```diff
@@ -2609,3 +2667,267 @@ StETofCalibMaker::writeHistograms()
+		
+	int EvtId = intVec.at(i) % 4000000000;   
+
+	startVec.push_back(EvtId);
+	mMasterStartVec.push_back(EvtId);
+	
+	lastEvtId = EvtId;
+
+	break;		
+		
+        // decode nonZero/stateChange ints ( int = 41.xxx.x00.000 = 1 states only)
+	case 41:
+		
+	int tmp       = intVec.at(i) % 4100000000;
+	int stateInt1 = tmp / 10000;
+	int Get4Id1 = -1;
+	int get4state1 = -1;
+	
+	if(stateInt1 <= 6912) {	 
+	  Get4Id1    = stateInt1 % 1728;
+	  get4state1 = stateInt1 / 1728;
+	}
+	
+	stateVec[lastEvtId].push_back(get4state1);
+	get4IdVec[lastEvtId].push_back(Get4Id1);
+
+	break;
+
+	default:
+	LOG_ERROR << "Get4 state not well defined -> Check db / state file !" << endm;	
+     }
+   }
+
+   // fill stateMap & steering vecs with EvtZero entries: read in first 1728 states & times
+   for(int i = 0; i< 1728;i++){
+     
+     for(unsigned int j=0; j< startVec.size(); j++){
+       
+       unsigned long int key = startVec.at(j);
+       
+       for(unsigned int n =0; n < get4IdVec.at(key).size(); n++){
+	 
+	 //steering vecs
+	 if(i == get4IdVec.at(key).at(n)){ 
+	   mStateVec[i].push_back(stateVec.at(key).at(n));
+	   mStartVec[i].push_back(startVec.at(j));	   
+	 } 	 
+       }  
+     }
+   }
+   
+   //set map validity check evtids ... EvtZero states only valid to first change of state on any get4
+   mStateMapStart = 0 ;
+   mStateMapStop  = startVec.at(0);
+   mDbEntryStart   = startVec.at(0);
+   mDbEntryStop    = startVec.at((startVec.size()-1));
+
+
+   if(fileZero){
+     mDbEntryStart   = 0;
+   }
+  
+   sort( mMasterStartVec.begin(), mMasterStartVec.end() );
+   mMasterStartVec.erase( unique( mMasterStartVec.begin(), mMasterStartVec.end() ), mMasterStartVec.end() );
+
+ }
+
+// -------------------------------------------------------------------------------
+
+void StETofCalibMaker::checkGet4State(unsigned long int eventNr){
+
+  if(eventNr >= mStateMapStart && eventNr < mStateMapStop) {
+    return; // stateMap still valid 
+  }
+
+  unsigned long int closestStop  = 99999999;
+  unsigned long int closestStart = 0;
+
+  //loop over stateMap
+
+  for(unsigned int i =0; i< 1728; i++){
+
+    std::vector<unsigned long int> tmpStart = mStartVec[i];
+    std::vector<short>             tmpState = mStateVec[i];
+
+    //find closest evtNr & state for each Get4
+    unsigned int      indexStart = 0;
+    // int               indexStop  = 0;
+    short             newState   = 0;
+    // unsigned long int newStart   = 0;
+    // unsigned long int newStop    = 0;
+
+    if(tmpStart.size() == 0 ) continue;
+    
+    auto lower = std::lower_bound(tmpStart.begin(), tmpStart.end(), eventNr);
+    indexStart = std::distance(tmpStart.begin(), lower);    
+    if(indexStart > 0) indexStart--;
+
+    //event past last change on get4 in entry
+    if(eventNr > tmpStart.at((tmpStart.size() -1))){
+      indexStart = (tmpStart.size() -1);
+    }
+  
+    if((indexStart < (tmpStart.size() -1 )) && eventNr == tmpStart.at(indexStart + 1)){
+	indexStart++;	
+    }
```

Comment:
This logic behind the index adjustment is very obscure. Please add a comment describing it in words or reconsider how the index is used.